### ISSUE TITLE
App crashes on folders with no image files at top level of folder structure

### ISSUE THREAD

**Issue by mikemcduffie** [pkrll/ComicZipper-2#14]
I've begun testing with lager groups of folders and larger sized folders and I'm seeing crashes. When processor subsequent large folders the app crashes overtime so far. Normally on the second large sized folder. 

I get the same exception each time:

Application Specific Information:
*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 9223372036854775807 beyond bounds [0 .. 0]'
terminating with uncaught exception of type NSException
abort() called

Archiving the same folder individually sometimes produces a complete uncorrupted archive but the app still crashes. A few times it crashes before completion leaving the zip temp file. I'm testing with 2 folders larger than 500MB and 1folder larger than 600MB. I can get the crash with any of the large sized folders. 

I can send complete crash reports if it would help. 

I'm running Mac OS X 10.10.5 and using version 2.3.4 as reported in the about box. I freshly downloaded the source and did a new build right before testing.

I don't think the built-in zip tool is the issue.  A simple applescript droplet with do shell command of:

            zip -r " & zipFile & " " & fileName & " -x '*/\.*' thumbs.db"

batch processes the same folders with no issues.

**Comment by mikemcduffie** [comment]
Crap. Forgot to mention this on 2.3.4

**Comment by pkrll** [comment]
I can't reproduce your error. I tested on 2 folders over 500MB, and 4 folders, between 700MB and 1,26GB.

**Comment by pkrll** [comment]
Did the folders have images inside of them?

**Comment by mikemcduffie** [comment]
Again testing on yesterday's version 2.3.4. 

Pretty standard (albeit larger omnibus/TPB) comic archive folders for testing. 

<pre><code>
Folder 1: 533,501,195 bytes (535.9 MB on disk) for 1,154 items [no subfolders]
Folder 2: 692,801,540 bytes (693.5 MB on disk) for 333 items [1 subfolder with all images in it; .DS_Store in top level folder]
Folder 3: 532,295,962 bytes (532.8 MB on disk) for 245 items [1 subfolder with all images in it; no files in top level folder]
</code></pre>

Interestingly, CZ2 shows 22kb for Folder 2 in the status despite a slightly lengthy delay in displaying the window. I've noticed the delay with folders/groups of folders. I assume it's from the size counting. '

*Question: why cumulative  size rather than number of folders/archive? I'm on a fairly fast SSD and I the delay until window display is noticeable for larger inputs.*

Folder 1 alone: compressed correctly and did not crash. Alert sound played as app was in foreground. 533.5MB cumulative size displayed (*different than Finder's cumulative size?*).

Folder 2 alone: produced a complete CBZ archive: 671,548,119 bytes (671.6 MB on disk)
and crashed:

<pre><code>Application Specific Information:
*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 9223372036854775807 beyond bounds [0 .. 1]'
terminating with uncaught exception of type NSException
abort() called
</code></pre>

A second test produced the same results (and the -1 archive filename). Same exception.

Folder 3 alone: shows 19kb (again counting just top level file sizes despite the lengthy launch delay?). Again same result (and same exception) as Folder 2 for two test iterations. 

Could not reproduce the crash before completion (leaving ZIP temp file) this morning.

Rerunning with just the subfolder for Folder 2 shows 692.8 MB for cum filesize. And ran without crash.

I suspect there is something with subfolders with all or almost all of the files below the first level...


Tested with all images at top level and empty subfolder: PASSED.

Tested with all but 15 of the images at the top and the rest in a subfolder: PASSED.

Tested with 150 (300 MB) of files moved to the subfolder: PASSED.

Tested with 15 images at the top and the remainder (658.3 MB) in the subfolder: PASSED.

Tested with 1 image only at top level: PASSED.

Tested will all images in the subfolder: CRASHED (same exception).

<pre><code>Application Specific Information:
*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 9223372036854775807 beyond bounds [0 .. 1]'
terminating with uncaught exception of type NSException
abort() called
</code></pre>

Tested with 0 length readme file at top level: CRASHED.

Tested with 14b readme file at top level: CRASHED.

Tested with 34kB readme file at top level: CRASHED (resulting archive had  readme file compressed to 100 bytes).

Tested with 3kB jpg image at top level: PASSED.

Tested with no files at top level and just 24 images (54.1 mB) in subfolder: CRASHED.

*Conclusion: If the resulting archive has no files or near 0 length file(s) at the top level after excluding files, the app crashes. Even a small uncompressable file at the top level will compress without crashing. Large archive/folder size is not the issue as a regular size archive crashes as well.*

Again a simple applescript droplet with `do shell` zip command will compress all the folders without error. 

Is the filesize computation needed? I'm sure you have a reason, but to me the number of archives to process and the status (e.g., "3 0f 8 completed") is more useful to me. Admittedly it'd be great to have a progress indicator of the currently active zip activity (especially for larger archives) but the unix tool does not provide progress feedback that I'm aware of. 

At the risk of too much input (let me know when you get tired of this), I was going to file an enhancement request to change icon badging to "items remaining to process" rather than total items. Personally, if I check the dock or switcher for status I prefer to see remaining items to complete. The Unarchiver uses this feedback if you want to see it in action.

I hope this helps.

**Comment by pkrll** [comment]
No, I appreciate the feedback. It's great having someone actually testing out the app, finding bugs and bad implementations).

If I understand you correctly, I think the app crashes when trying to compress a folder with no images at the top level. As you say, the zip tool does its job, the problem is that the app after completion tries to retrieve the first image of the folder, without checking if it exists. It's something I should have included, but I've never even thought of compressing subfolders :) It's a simple fix, though.

About the file size computation: I agree about the status indicator (3 of 8 completed) as a better use of the top space than the total file size. I'd love a real progress indicator too, but, the zip tool does not make it possible.

Good point about the badge icon, I'll create a new issue for this and the status indicator (instead of total file size).

**Comment by mikemcduffie** [comment]
Yes, it crashes without an image in the top level. I had (wrongly) assumed it was something to do with file size computations or something similar. 

Are you setting an icon with a top level image, or some other action? Otherwise why do anything with the image files?

*Note: I really need to start reviewing the code to see what is going on. ;)*

**Comment by pkrll** [comment]
Yes, the image will replace the folder icon on processed items.

A framework seems to much, but maybe in the future.

**Comment by mikemcduffie** [comment]
Fixed for me in the 9/27 build.

**Comment by pkrll** [comment]
Great.

**Comment by mikemcduffie** [comment]
Did a little research on a possible framework implementation. Understand this may be a bridge too far, but wanted to share what I found and, if possible, get your thoughts on whether the implementation would resolve the progress feedback issue.

I found a few frameworks and wrappers, but ZipArchive seems to have the most promise.
   https://github.com/ZipArchive/ZipArchive

It includes a method for progress (zipArchiveProgressEvent).

While unzipping appears to require counting files and calculating progress, would this report progress for zipping a group (folder) of files?
   https://github.com/ZipArchive/ZipArchive/issues/118
